src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -0,0 +1,254 @@
+	r.tags = firstEntry.tags
+
+	// use empty var in inner loop with "for i := range" to have compiler use memclr optimization
+	// see: https://codereview.appspot.com/137880043
+	var emptyRecord BlockRecord
+	for i := range r.records {
+		r.records[i] = emptyRecord
+	}
+	r.records = r.records[:0]
+	r.records = append(r.records, BlockRecord{firstEntry.data, firstEntry.checksum})
+
+	for len(r.minHeap) > 0 && r.minHeap[0].id.Equal(firstEntry.id) {
+		nextEntry, err := r.readOne()
+		if err != nil {
+			// Close the resources that were already read but not returned to the consumer:
+			r.id.Finalize()
+			r.tags.Close()
+			for _, record := range r.records {
+				record.Data.DecRef()
+				record.Data.Finalize()
+			}
+			for i := range r.records {
+				r.records[i] = emptyRecord
+			}
+			r.records = r.records[:0]
+			r.err = err
+			return false
+		}
+		nextEntry.id.Finalize()
+		nextEntry.tags.Close()
+		r.records = append(r.records, BlockRecord{nextEntry.data, nextEntry.checksum})
+	}
+
+	return true
+}
+
+func (r *crossBlockReader) Current() (ident.ID, ident.TagIterator, []BlockRecord) {
+	return r.id, r.tags, r.records
+}
+
+func (r *crossBlockReader) readOne() (*minHeapEntry, error) {
+	if len(r.minHeap) == 0 {
+		return nil, errEmptyReader
+	}
+
+	entry := heap.Pop(&r.minHeap).(*minHeapEntry)
+	if r.dataFileSetReaders[entry.dataFileSetReaderIndex] != nil {
+		nextEntry, err := r.readFromDataFileSet(entry.dataFileSetReaderIndex)
+		if err == io.EOF {
+			// will no longer read from this one
+			r.dataFileSetReaders[entry.dataFileSetReaderIndex] = nil
+		} else if err != nil {
+			return nil, err
+		} else {
+			heap.Push(&r.minHeap, nextEntry)
+		}
+	}
+
+	return entry, nil
+}
+
+func (r *crossBlockReader) init() error {
+	r.initialized = true
+	r.minHeap = make([]*minHeapEntry, 0, len(r.dataFileSetReaders))
+
+	for i := range r.dataFileSetReaders {
+		entry, err := r.readFromDataFileSet(i)
+		if err == io.EOF {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		r.minHeap = append(r.minHeap, entry)
+	}
+
+	heap.Init(&r.minHeap)
+
+	return nil
+}
+
+func (r *crossBlockReader) readFromDataFileSet(index int) (*minHeapEntry, error) {
+	id, tags, data, checksum, err := r.dataFileSetReaders[index].Read()
+
+	if err == io.EOF {
+		return nil, err
+	}
+
+	if err != nil {
+		multiErr := xerrors.NewMultiError().
+			Add(err).
+			Add(r.Close())
+		return nil, multiErr.FinalError()
+	}
+
+	return &minHeapEntry{
+		dataFileSetReaderIndex: index,
+		id:                     id,
+		tags:                   tags,
+		data:                   data,
+		checksum:               checksum,
+	}, nil
+}
+
+func (r *crossBlockReader) Err() error {
+	return r.err
+}
+
+func (r *crossBlockReader) Close() error {
+	// Close the resources that were buffered in minHeap:
+	for i, entry := range r.minHeap {
+		entry.id.Finalize()
+		entry.tags.Close()
+		entry.data.DecRef()
+		entry.data.Finalize()
+		r.minHeap[i] = nil
+	}
+
+	r.minHeap = r.minHeap[:0]
+	return nil
```

Comment:
Hmm, not sure. The consumer might still be referencing these. OTOH, if the consumer called `Next` and did not call `Current`, this will leak. Looks like the issue here arises because of the switch from `Read` to `Next`/`Current` semantics. Are you aware of any practice in handling similar situations?

Reply:
Actually I think it's fine to have as is, with these getting closed the correct usage pattern for iterators we have is expected to be that your data is only valid until you call `Next()` again, generally of the form:

```golang
for it.Next() {
 curr := it.Current()
 applySomeFunction(curr)
}

if err := it.Err(); err != nil { // report error }
it.Close()
```

I was misreading the details in the Next() call and where you were closing these at the moment, makes sense that it's up to the clients to deal with closing it here as well :+1: